From f2c4975451756c825ef2b6bf7ae8249587bb0fc5 Mon Sep 17 00:00:00 2001
From: Huacai Chen <chenhc@lemote.com>
Date: Thu, 1 Dec 2016 09:51:35 +0800
Subject: [PATCH 177/274] AOSCOS: MIPS: Loongson 3: Introduce
 disable_unused_cpus() to save power

When using maxcpus=n to limit the number of cores, the initial offlined
cores are running a infinite loop in firmware and their clocks are not
disabled. For power saving, we take the unused cores up and then down
at boot time. When resume from STR, we do the same.

[Mingcong Bai: Resolved merge conflicts in
arch/mips/include/asm/mach-loongson64/loongson.h,
arch/mips/loongson64/pm.c. Also, apply minor lint on grammar in the
original commit message.]

Signed-off-by: Huacai Chen <chenhc@lemote.com>
Signed-off-by: Hongliang Tao <taohl@lemote.com>
Signed-off-by: Hua Yan <yanh@lemote.com>
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>

Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 .../include/asm/mach-loongson64/loongson.h     |  6 ++++++
 arch/mips/loongson64/pm.c                      |  3 +++
 arch/mips/loongson64/smp.c                     | 18 ++++++++++++++++++
 3 files changed, 27 insertions(+)

diff --git a/arch/mips/include/asm/mach-loongson64/loongson.h b/arch/mips/include/asm/mach-loongson64/loongson.h
index 70399b6cd2ae..a8eb17134e14 100644
--- a/arch/mips/include/asm/mach-loongson64/loongson.h
+++ b/arch/mips/include/asm/mach-loongson64/loongson.h
@@ -268,4 +268,10 @@ extern u64 loongson_freqctrl[MAX_PACKAGES];
 extern struct cpufreq_frequency_table gs464_clockmod_table[];
 #endif
 
+#ifdef CONFIG_HOTPLUG_CPU
+extern int disable_unused_cpus(void);
+#else
+static inline int disable_unused_cpus(void) { return 0; }
+#endif
+
 #endif /* __ASM_MACH_LOONGSON64_LOONGSON_H */
diff --git a/arch/mips/loongson64/pm.c b/arch/mips/loongson64/pm.c
index b750094c57c5..c65a39cd7948 100644
--- a/arch/mips/loongson64/pm.c
+++ b/arch/mips/loongson64/pm.c
@@ -35,6 +35,9 @@ static int lefi_pm_enter(suspend_state_t state)
 
 static void lefi_pm_wake(void)
 {
+#ifdef CONFIG_CPU_LOONGSON3
+	disable_unused_cpus();
+#endif
 #ifdef CONFIG_LOONGSON64_CPUAUTOPLUG
 	extern int autoplug_enabled;
 	autoplug_enabled = cached_autoplug_enabled;
diff --git a/arch/mips/loongson64/smp.c b/arch/mips/loongson64/smp.c
index f8e2fb4c7cd8..f0847ef7fc8d 100644
--- a/arch/mips/loongson64/smp.c
+++ b/arch/mips/loongson64/smp.c
@@ -841,6 +841,24 @@ static int register_loongson3_notifier(void)
 }
 early_initcall(register_loongson3_notifier);
 
+int disable_unused_cpus(void)
+{
+	int cpu;
+	struct cpumask tmp;
+
+	cpumask_complement(&tmp, cpu_online_mask);
+	cpumask_and(&tmp, &tmp, cpu_possible_mask);
+
+	for_each_cpu(cpu, &tmp)
+		cpu_up(cpu);
+
+	for_each_cpu(cpu, &tmp)
+		cpu_down(cpu);
+
+	return 0;
+}
+core_initcall(disable_unused_cpus);
+
 #endif
 
 const struct plat_smp_ops loongson3_smp_ops = {
-- 
2.47.1

